src/control/server/instance_storage_rpc.go

```diff
@@ -340,12 +341,16 @@ func scanEngineBdevsOverDrpc(ctx context.Context, engine Engine, pbReq *ctlpb.Sc
 	}
 	seenCtrlrs := make(map[string]*ctlpb.NvmeController)
 
-	for _, sd := range scanSmdResp.Devices {
+	for i, sd := range scanSmdResp.Devices {
 		if sd.Ctrlr == nil {
 			return nil, errors.Errorf("smd %q has no ctrlr ref", sd.Uuid)
 		}
 
 		addr := sd.Ctrlr.PciAddr
+		if addr == "" {
+			// Mock identifier for emulated NVMe mode where devices have no PCI-address.
+			addr = fmt.Sprintf("0000:00:0.%d", i)
```

Comment:
My only question is whether this could possibly conflict with a real device, since it is used in the `seenCtrlrs` map. I wonder if it would be better to use a "name" which could be either a PCI address for the real device, or the file name for emulated nvme.

Reply:
In DPDK, `PCI_ANY_ID` is defined as `0xffff`: https://doc.dpdk.org/api/rte__pci_8h.html#a53aca768a081fcf56089353d805ab77c

That also seems to match the kernel: https://elixir.bootlin.com/linux/v6.12.1/source/include/linux/mod_devicetable.h#L18

Might be better to use that value, as there is zero possibility of conflicting with a real vendor ID that way?

Reply:
so are you suggesting `FFFF:00:0.%d` as the format string?

Reply:
I think so, if the goal is to avoid any conflict with a device that might be found via the topology scan. I can't find an official PCI spec to cite, but from what I found about how linux and DPDK do things, that value is a "can't happen" value, i.e. there's no way that a valid device could show up with that vendor ID.

Reply:
aren't we getting confused here between PCI-address and PCI-ID? https://wiki.debian.org/HowToIdentifyADevice/PCI

Reply:
the first segment of the PCI address is the domain not the vendor ID afais

Reply:
I've improved the resilience against collisions by using `0xffff` PCI-domain segment and `0xf` func value (neither would be used in a "real" NVMe device address as domain segment is either `0x10000` or `0x0000` in almost all the addresses and BDF func value is 0-7).